### DISCUSSION TITLE
用 Google Apps Script 搭配 Google Sheet 開一個簡易 API

### DISCUSSION THREAD

**Discussion by JSHT** [1101-GS4719-C/Python-Community-GS4719-C#546] [General]
# Google Apps Script
之前有人提到了可以使用 Postman 來 call API。這邊有一個可以自己建立簡易 API 的方法，這邊具體介紹及教學可以參考 [用 Google Apps Script 搭配 Google 試算表開一個 API](https://israynotarray.com/other/20220211/2705866340/)。

### google sheet
按照文章中建立 google sheet 並按照對應格式放資料
EX.
![image](https://user-images.githubusercontent.com/48956859/161483828-834ac40c-e083-4d69-9da2-af0491ec31d7.png)

### Example code
```=javascript
function doGet(){
  // 放入自己的 google sheet ID
  const sheetID = "1UA4OZEjbomWC7lfRLBbhHhoGAEN8Ca_rEuar_3FFlbE"
  const sheetName = '工作表1'

  const ss = SpreadsheetApp.openById(sheetID);

  const sheet = ss.getSheetByName(sheetName);
  // const sheet = ss.getSheets()[0];

  // raw data, not json type
  const values = sheet.getDataRange().getValues();
  const keys = values[0];
  console.log(keys);
  values.shift();
  
  let data = [];
  values.forEach((item) => {
    let obj = {};
    keys.forEach((key, index)=>{
      obj[key] = item[index];
    })
    data.push(obj);
  })

  console.log(data);
  // console.log(SheetName.getDataRange().getValues());

  const dataExportJSON = JSON.stringify(data)
  return ContentService.createTextOutput(dataExportJSON).setMimeType(ContentService.MimeType.JSON); 
}
```

## 使用 Postman call API
[API 網址](https://script.google.com/macros/s/AKfycbxEPbh7jSHZ4y0t6JyaQP-WlEq4sszXdvqbbIOXLunnPKSWI6NskWSBUqYqXjtreVjj/exec)
![image](https://user-images.githubusercontent.com/48956859/161483600-389eddaa-ac8a-4fde-8477-5d47b6f130e1.png)

## 結語
Google App Scripts 是個輕便的功能，除了能夠製作簡易的 API (get，post) 之外，也有其他各項功能。網路上的教學也有很多，可以自己動手玩玩看。

**Comment by JSHT** [comment]
這邊嘗試了一個簡單的應用範例 [Codepen](https://codepen.io/jsht-the-animator/pen/gOooMxj?editors=0011)

**Reply by chia-kai-chang** [reply]
不錯耶，你會使用Javascript

**Comment by chia-kai-chang** [comment]
也可以用 [Flask](https://github.com/NCU-GS4719-Python/Python-Community/tree/main/Python_Flask) 做一個後端API網站，然後把網站部署到Azure WebApp，也可以喔。

**Reply by JSHT** [reply]
了解，謝謝老師

**Comment by hapiwangy** [comment]
是一個沒有接觸過的技術，謝謝樓主分享
